File: src/main/javabak/item/WrenchItem.java
package redcrafter07.processed.item;

import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.block.DirectionalBlock;
import net.minecraft.world.level.block.HorizontalDirectionalBlock;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.DirectionProperty;
import redcrafter07.processed.block.WrenchInteractableBlock;

import javax.annotation.Nullable;
import java.util.List;

import static redcrafter07.processed.item.ModDataComponents.WRENCH_MODE;

public class WrenchItem extends ModItem {
    public WrenchItem() {
        super(new Properties().stacksTo(1), "wrench");
    }

    public static WrenchMode getMode(ItemStack stack) {
        return stack.getOrDefault(WRENCH_MODE, WrenchMode.Config);
    }

    public static ItemStack setMode(ItemStack stack, WrenchMode mode) {
        stack.set(WRENCH_MODE, mode);
        return stack;
    }

    private static final List<DirectionProperty> facingProperties = List.of(
            DirectionalBlock.FACING,
            HorizontalDirectionalBlock.FACING,
            BlockStateProperties.HORIZONTAL_FACING,
            BlockStateProperties.FACING
    );

    @Override
    public ItemStack getDefaultInstance() {
        return setMode(super.getDefaultInstance(), WrenchMode.Config);
    }

    @Override
    public InteractionResult useOn(UseOnContext context) {
        Player player = context.getPlayer();
        if (player == null) return InteractionResult.PASS;
        ItemStack stack = player.getItemInHand(context.getHand());
        if (stack.getItem() instanceof WrenchItem item) {
            WrenchMode wrenchMode = getMode(stack);

            return switch (wrenchMode) {
                case WrenchMode.Config -> configure(stack, context);
                case WrenchMode.Rotate -> rotate(stack, context);
            };
        }
        return InteractionResult.PASS;
    }

    private InteractionResult rotate(ItemStack stack, UseOnContext context) {
        var blockState = context.getLevel().getBlockState(context.getClickedPos());
        var player = context.getPlayer();
        if (player == null) return InteractionResult.PASS;

        for (var property : facingProperties) {
            if (blockState.hasProperty(property)) {
                var facing = blockState.getValue(property);
                var newFacing = (player.isShiftKeyDown()) ? facing.getCounterClockWise() : facing.getClockWise();
                var newState = blockState.setValue(property, newFacing);
                context.getLevel().setBlock(context.getClickedPos(), newState, 3);
                return InteractionResult.SUCCESS;
            }
        }

        return InteractionResult.PASS;
    }

    private InteractionResult configure(ItemStack stack, UseOnContext context) {
        var blockState = context.getLevel().getBlockState(context.getClickedPos());

        if (blockState.getBlock() instanceof WrenchInteractableBlock block) {
            block.onWrenchUse(context, blockState);
            return InteractionResult.SUCCESS;
        }

        if (context.getLevel().getBlockEntity(context.getClickedPos()) instanceof WrenchInteractableBlock blockEntity) {
            blockEntity.onWrenchUse(context, blockState);
            return InteractionResult.SUCCESS;
        }

        return InteractionResult.PASS;
    }

    @Override
    public @Nullable MutableComponent

    getAdditionalTooltip(ItemStack stack, TooltipContext context, TooltipFlag flag) {
        return Component.translatable(
                "item.processed.wrench.mode", Component.translatable(getMode(stack).translation())
        );
    }
}

File: src/main/kotlin/redcrafter07/processed/items/WrenchItem.kt
package redcrafter07.processed.items

import net.minecraft.core.Direction
import net.minecraft.network.chat.MutableComponent
import net.minecraft.world.InteractionResult
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.TooltipFlag
import net.minecraft.world.item.context.UseOnContext
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import redcrafter07.processed.Translations
import redcrafter07.processed.block.WrenchInteractableBlock

class WrenchItem : ModItem(Properties().stacksTo(1), "wrench") {
    companion object {
        fun getMode(stack: ItemStack): WrenchMode = stack.getOrDefault(ModDataComponents.WRENCH_MODE, WrenchMode.Config)

        fun setMode(stack: ItemStack, mode: WrenchMode): ItemStack {
            stack.set(ModDataComponents.WRENCH_MODE, mode)
            return stack
        }
    }

    override fun getDefaultInstance(): ItemStack = setMode(super.defaultInstance, WrenchMode.Config)

    override fun useOn(context: UseOnContext): InteractionResult {
        val player = context.player ?: return InteractionResult.PASS
        val stack = player.getItemInHand(context.hand)
        if (stack.item is WrenchItem) {
            val mode = getMode(stack)

            return when (mode) {
                WrenchMode.Rotate -> rotate(context)
                WrenchMode.Config -> configure(context)
            }
        }

        return super.useOn(context)
    }


    private fun rotate(context: UseOnContext): InteractionResult {
        val blockState = context.level.getBlockState(context.clickedPos)
        val player = context.player ?: return InteractionResult.PASS

        if (blockState.hasProperty(BlockStateProperties.FACING)) {
            val facing = blockState.getValue(BlockStateProperties.FACING)
            val newData = facing.get3DDataValue() + if (player.isShiftKeyDown) -1 else 1
            val newFacing = Direction.from3DDataValue(if (newData > 5) 0 else if (newData < 0) 5 else newData)
            val newState = blockState.setValue(BlockStateProperties.FACING, newFacing)
            context.level.setBlock(context.clickedPos, newState, 3)
        } else if (blockState.hasProperty(BlockStateProperties.HORIZONTAL_FACING)) {
            val facing = blockState.getValue(BlockStateProperties.HORIZONTAL_FACING)
            val newFacing = if (player.isShiftKeyDown) facing.counterClockWise else facing.clockWise
            val newState = blockState.setValue(BlockStateProperties.HORIZONTAL_FACING, newFacing)
            context.level.setBlock(context.clickedPos, newState, 3)
        } else return InteractionResult.PASS

        return InteractionResult.SUCCESS
    }

    private fun configure(context: UseOnContext): InteractionResult {
        val blockState = context.level.getBlockState(context.clickedPos)

        val block = blockState.block
        if (block is WrenchInteractableBlock) {
            block.onWrenchUse(context, blockState)
            return InteractionResult.SUCCESS
        }

        val blockEntity = context.level.getBlockEntity(context.clickedPos)
        if (blockEntity is WrenchInteractableBlock) {
            blockEntity.onWrenchUse(context, blockState)
            return InteractionResult.SUCCESS
        }

        return InteractionResult.PASS
    }

    override fun getAdditionalTooltip(
        stack: ItemStack, context: TooltipContext, flag: TooltipFlag
    ): List<MutableComponent> = listOf(Translations.wrenchModeTooltip(getMode(stack)))
}